modules/calc/src/main/java/com/opengamma/strata/calc/marketdata/DefaultMarketDataFactory.java

```diff
@@ -336,62 +317,22 @@ private void addValue(
   }
 
   /**
-   * Builds items of observable market data.
+   * Provides observable market data.
    *
-   * @param ids  IDs of the market data that should be built
+   * @param ids  the market data identifiers that are required
    */
-  private Map<ObservableId, Result<Double>> buildObservableData(Set<ObservableId> ids) {
-    // We need to convert between the input IDs from the requirements and the source IDs
-    // which are passed to the builder and used to request the data.
-    Map<ObservableId, ObservableId> observableIdToRequirementId = new HashMap<>();
-    // IDs that are in the requirements but have no mapping to an ID the data provider understands
-    Set<ObservableId> unmappedIds = new HashSet<>();
-
-    // TODO Mapping of IDs should probably go inside the ObservableMarketDataBuilder
-    for (ObservableId id : ids) {
-      Optional<ObservableId> obsSourceId = observableIdMapping.lookupId(id);
-
-      if (obsSourceId.isPresent()) {
-        observableIdToRequirementId.put(obsSourceId.get(), id);
-      } else {
-        unmappedIds.add(id);
-      }
-    }
-    Map<ObservableId, Result<Double>> builtValues = observablesBuilder.build(observableIdToRequirementId.keySet());
-    ImmutableMap.Builder<ObservableId, Result<Double>> builder = ImmutableMap.builder();
-    // Put the built data into the results, mapping the source ID to the ID that was passed in
-    builtValues.keySet().stream().forEach(id -> builder.put(observableIdToRequirementId.get(id), builtValues.get(id)));
-    // Add failures for IDs that don't have mappings to market data source IDs
-    unmappedIds.forEach(id -> builder.put(id, noMappingResult(id)));
-    return builder.build();
-  }
-
-  /**
-   * Returns a failure result for an observable ID that can't be mapped to an ID
-   * recognized by the market data source.
-   *
-   * @param id  an observable ID that can't be mapped to an ID recognized by the market data source
-   * @return a failure result for the ID
-   */
-  private <T> Result<T> noMappingResult(ObservableId id) {
-    return Result.failure(FailureReason.MISSING_DATA, "No source ID mapping found for ID {}", id);
+  private Map<ObservableId, Result<Double>> provideObservableData(Set<ObservableId> ids) {
+    Map<ObservableId, Result<Double>> builtValues = observableDataProvider.provideObservableData(ids);
+    return ids.stream()
```

Comment:
I think this logic should be in the `ObservableDataProvider`. It should guarantee to return a result for every ID for which data is requested. If the data isn't available it should return a failure.